File: Solution/JavaScript/The-Magical-Forest-of-Algora.js
const DanceMove = {
    Twirl: 0,
    Leap: 1,
    Spin: 2,
    Slide: 3,
  };
  
  class Creature {
    constructor(name, danceMoves) {
      this.name = name;
      this.danceMoves = danceMoves;
    }
  }
  
  const danceMoves = {
    [DanceMove.Twirl]: "Twirl",
    [DanceMove.Leap]: "Leap",
    [DanceMove.Spin]: "Spin",
    [DanceMove.Slide]: "Slide",
  };
  
  const effects = {
    '0,0,0': "A flurry of autumn leaves fills the air.",
    '0,0,1': "A soft mist envelops the forest.",
    '0,0,2': "A chorus of crickets begins to play.",
    '0,0,3': "Fireflies light up the forest.",
    '0,1,0': "A gentle breeze rustles the leaves.",
    '0,1,1': "A rainbow forms in the morning dew.",
    '0,1,2': "A waterfall appears from nowhere.",
    '0,1,3': "A flock of birds takes flight.",
    '0,2,0': "The forest floor blooms with wildflowers.",
    '0,2,1': "A gentle rain starts falling.",
    '0,2,2': "A swarm of butterflies fills the air.",
    '0,2,3': "A stream begins to babble in harmony with the music.",
    '0,3,0': "The air fills with the scent of fresh flowers.",
    '0,3,1': "A brilliant light illuminates the forest.",
    '0,3,2': "The forest glows with a soft light.",
    '0,3,3': "A chorus of forest creatures begins to sing.",
    '1,0,0': "A soft glow envelops the forest.",
    '1,0,1': "A symphony of bird songs fills the air.",
    '1,0,2': "A cascade of petals falls from the trees.",
    '1,0,3': "A gentle snow begins to fall.",
    '1,1,0': "A rainbow appears in the sky.",
    '1,1,1': "A waterfall springs from the ground.",
    '1,1,2': "A gust of wind blows through the forest.",
    '1,1,3': "A chorus of frogs begins to sing.",
    '1,2,0': "A swarm of bees buzzes in harmony with the music.",
    '1,2,1': "A field of flowers blooms instantly.",
    '1,2,2': "A group of squirrels performs a synchronized dance.",
    '1,2,3': "A group of fireflies begins a light show.",
    '1,3,0': "A group of mushrooms glow brightly.",
    '1,3,1': "A group of trees sway in time with the music.",
    '1,3,2': "A group of rocks levitate in the air.",
    '1,3,3': "A group of clouds forms a beautiful pattern in the sky.",
    '2,0,0': "A constellation of stars appears in the sky.",
    '2,0,1': "A sudden gust of wind lifts leaves into the air.",
    '2,0,2': "A group of rabbits perform a synchronized hop.",
    '2,0,3': "A sudden bloom of flowers covers the forest floor.",
    '2,1,0': "A group of fish leap out of a stream in a beautiful arc.",
    '2,1,1': "A group of trees bend and sway to the music.",
    '2,1,2': "A group of clouds change shape to form a beautiful scene.",
    '2,1,3': "A group of fireflies spell out a message in the air.",
    '2,2,0': "A group of birds fly in a mesmerizing pattern.",
    '2,2,1': "A group of butterflies flutter in a beautiful dance.",
    '2,2,2': "A group of deer prance in a circle.",
    '2,2,3': "A group of frogs leap in a synchronized dance.",
    '2,3,0': "A group of mushrooms suddenly grow and emit a soft glow.",
    '2,3,1': "A group of trees suddenly grow fruit.",
    '2,3,2': "A group of flowers bloom and emit a beautiful fragrance.",
    '2,3,3': "A group of rocks suddenly sparkle with gemstones.",
    '3,0,0': "A group of trees suddenly grow taller.",
    '3,0,1': "A group of flowers suddenly change color.",
    '3,0,2': "A group of rocks suddenly become transparent.",
    '3,0,3': "A group of clouds suddenly take on a beautiful color.",
    '3,1,0': "A group of birds suddenly start singing in harmony.",
    '3,1,1': "A group of butterflies suddenly start flying in a beautiful pattern.",
    '3,1,2': "A group of deer suddenly start prancing in a circle.",
    '3,1,3': "A group of frogs suddenly start leaping in a synchronized dance.",
    '3,2,0': "A group of mushrooms suddenly start glowing in different colors.",
    '3,2,1': "A group of trees suddenly start swaying in time with the music.",
    '3,2,2': "A group of flowers suddenly start blooming and emitting a beautiful fragrance.",
    '3,2,3': "A group of rocks suddenly start sparkling with different colored gemstones.",
    '3,3,0': "A group of clouds suddenly start forming different shapes.",
    '3,3,1': "A group of birds suddenly start flying in a mesmerizing pattern.",
    '3,3,2': "A group of butterflies suddenly start fluttering in a beautiful dance.",
    '3,3,3': "A group of frogs suddenly start leaping in a synchronized dance.",
  };
  
  const lox = new Creature("Lox", [DanceMove.Twirl, DanceMove.Leap, DanceMove.Spin, DanceMove.Twirl, DanceMove.Leap]);
  const drako = new Creature("Drako", [DanceMove.Spin, DanceMove.Twirl, DanceMove.Leap, DanceMove.Leap, DanceMove.Spin]);
  const flix = new Creature("Flix", [DanceMove.Slide, DanceMove.Spin, DanceMove.Twirl, DanceMove.Slide, DanceMove.Leap]);
  
  for (let i = 0; i < 5; i++) {
    const loxMove = lox.danceMoves[i];
    const drakoMove = drako.danceMoves[i];
    const flixMove = flix.danceMoves[i];
  
    printMove(lox.name, loxMove);
    printMove(drako.name, drakoMove);
    printMove(flix.name, flixMove);
  
    console.log();
  
    const effect = effects[`${loxMove},${drakoMove},${flixMove}`];
    if (effect) {
      console.log(`✨ ${effect} ✨`);
    } else {
      console.log("A magical effect happens, but it's hard to describe.");
    }
  
    console.log();
  }
  
  function printMove(name, move) {
    let moveName = danceMoves[move];
    if (!moveName) {
      moveName = "Unknown";
    }
  
    const red = "\x1b[31m";
    const yellow = "\x1b[33m";
    const reset = "\x1b[0m";
  
    console.log(`${red}${name}${reset} performs a ${yellow}${moveName}${reset}, `);
  }

File: Solution/Go/The-Magical-Forest-of-Algora.go
package main

import (
	"fmt"
)

type DanceMove int

const (
	Twirl DanceMove = iota
	Leap
	Spin
	Slide
)

type Creature struct {
	Name       string
	DanceMoves []DanceMove
}

type DanceMoveKey struct {
	First  DanceMove
	Second DanceMove
	Third  DanceMove
}

func main() {
	effects := map[DanceMoveKey]string{
		{Twirl, Twirl, Twirl}: "A flurry of autumn leaves fills the air.",
		{Twirl, Twirl, Leap}:  "A soft mist envelops the forest.",
		{Twirl, Twirl, Spin}:  "A chorus of crickets begins to play.",
		{Twirl, Twirl, Slide}: "Fireflies light up the forest.",
		{Twirl, Leap, Twirl}:  "A gentle breeze rustles the leaves.",
		{Twirl, Leap, Leap}:   "A rainbow forms in the morning dew.",
		{Twirl, Leap, Spin}:   "A waterfall appears from nowhere.",
		{Twirl, Leap, Slide}:  "A flock of birds takes flight.",
		{Twirl, Spin, Twirl}:  "The forest floor blooms with wildflowers.",
		{Twirl, Spin, Leap}:   "A gentle rain starts falling.",
		{Twirl, Spin, Spin}:   "A swarm of butterflies fills the air.",
		{Twirl, Spin, Slide}:  "A stream begins to babble in harmony with the music.",
		{Twirl, Slide, Twirl}: "The air fills with the scent of fresh flowers.",
		{Twirl, Slide, Leap}:  "A brilliant light illuminates the forest.",
		{Twirl, Slide, Spin}:  "The forest glows with a soft light.",
		{Twirl, Slide, Slide}: "A chorus of forest creatures begins to sing.",
		{Leap, Twirl, Twirl}:  "A soft glow envelops the forest.",
		{Leap, Twirl, Leap}:   "A symphony of bird songs fills the air.",
		{Leap, Twirl, Spin}:   "A cascade of petals falls from the trees.",
		{Leap, Twirl, Slide}:  "A gentle snow begins to fall.",
		{Leap, Leap, Twirl}:   "A rainbow appears in the sky.",
		{Leap, Leap, Leap}:    "A waterfall springs from the ground.",
		{Leap, Leap, Spin}:    "A gust of wind blows through the forest.",
		{Leap, Leap, Slide}:   "A chorus of frogs begins to sing.",
		{Leap, Spin, Twirl}:   "A swarm of bees buzzes in harmony with the music.",
		{Leap, Spin, Leap}:    "A field of flowers blooms instantly.",
		{Leap, Spin, Spin}:    "A group of squirrels performs a synchronized dance.",
		{Leap, Spin, Slide}:   "A group of fireflies begins a light show.",
		{Leap, Slide, Twirl}:  "A group of mushrooms glow brightly.",
		{Leap, Slide, Leap}:   "A group of trees sway in time with the music.",
		{Leap, Slide, Spin}:   "A group of rocks levitate in the air.",
		{Leap, Slide, Slide}:  "A group of clouds forms a beautiful pattern in the sky.",
		{Spin, Twirl, Twirl}:  "A constellation of stars appears in the sky.",
		{Spin, Twirl, Leap}:   "A sudden gust of wind lifts leaves into the air.",
		{Spin, Twirl, Spin}:   "A group of rabbits perform a synchronized hop.",
		{Spin, Twirl, Slide}:  "A sudden bloom of flowers covers the forest floor.",
		{Spin, Leap, Twirl}:   "A group of fish leap out of a stream in a beautiful arc.",
		{Spin, Leap, Leap}:    "A group of trees bend and sway to the music.",
		{Spin, Leap, Spin}:    "A group of clouds change shape to form a beautiful scene.",
		{Spin, Leap, Slide}:   "A group of fireflies spell out a message in the air.",
		{Spin, Spin, Twirl}:   "A group of birds fly in a mesmerizing pattern.",
		{Spin, Spin, Leap}:    "A group of butterflies flutter in a beautiful dance.",
		{Spin, Spin, Spin}:    "A group of deer prance in a circle.",
		{Spin, Spin, Slide}:   "A group of frogs leap in a synchronized dance.",
		{Spin, Slide, Twirl}:  "A group of mushrooms suddenly grow and emit a soft glow.",
		{Spin, Slide, Leap}:   "A group of trees suddenly grow fruit.",
		{Spin, Slide, Spin}:   "A group of flowers bloom and emit a beautiful fragrance.",
		{Spin, Slide, Slide}:  "A group of rocks suddenly sparkle with gemstones.",
	}

	lox := Creature{
		Name:       "Lox",
		DanceMoves: []DanceMove{Twirl, Leap, Spin, Twirl, Leap},
	}

	drako := Creature{
		Name:       "Drako",
		DanceMoves: []DanceMove{Spin, Twirl, Leap, Leap, Spin},
	}

	flix := Creature{
		Name:       "Flix",
		DanceMoves: []DanceMove{Slide, Spin, Twirl, Slide, Leap},
	}

	for i := 0; i < 5; i++ {
		loxMove := lox.DanceMoves[i]
		drakoMove := drako.DanceMoves[i]
		flixMove := flix.DanceMoves[i]

		printMove(lox.Name, loxMove)
		printMove(drako.Name, drakoMove)
		printMove(flix.Name, flixMove)

		fmt.Println()

		effect, ok := effects[DanceMoveKey{loxMove, drakoMove, flixMove}]
		if ok {
			fmt.Printf("✨ %s ✨\n", effect)
		} else {
			fmt.Println("A magical effect happens, but it's hard to describe.")
		}

		fmt.Println()
	}
}

var danceMoves = map[DanceMove]string{
	Twirl: "Twirl",
	Leap:  "Leap",
	Spin:  "Spin",
	Slide: "Slide",
}

func printMove(name string, move DanceMove) {
	moveName := danceMoves[move]
	if moveName == "" {
		moveName = "Unknown"
	}

	red := "\033[31m"
	yellow := "\033[33m"
	reset := "\033[0m"

	fmt.Printf("%s%s%s performs a %s%s%s, ", red, name, reset, yellow, moveName, reset)
}
